### ISSUE TITLE
run_forever()'s restart condition never met

### ISSUE THREAD

**Issue by Chrisrdouglas** [alpacahq/alpaca-trade-api-python#407]
in this loop on line 208 of stream.py, self._running is never set to false in the exception. retries starts at 0, then incremented in the exception. This has it compare 1>3 which is false. This means that self._running is never set to false. Then in finally block on line 226, the break statement on line 228 will never be reached. Wouldn't this stop it from restarting on the next pass of the loop?
```python
        while True: # stream.py - line 208
            try:
                if not self._running:
                    await self._start_ws()
                    self._running = True
                    retries = 0
                    await self._consume()
            except websockets.WebSocketException as wse:
                retries += 1
                if retries > int(os.environ.get('APCA_RETRY_MAX', 3)):
                    await self.close()
                    self._running = False
                    raise ConnectionError("max retries exceeded")
                if retries > 1:
                    await asyncio.sleep(
                        int(os.environ.get('APCA_RETRY_WAIT', 3)))
                log.warn('websocket error, restarting connection: ' +
                         str(wse))
            finally:
                print(retries)
                if not self._running:
                    break
                await asyncio.sleep(0.01)```

**Comment by Chrisrdouglas** [comment]
in this loop on line 208 of stream.py self._running is never set to false in the exception
`        while True:
            try:
                if not self._running:
                    await self._start_ws()
                    self._running = True
                    retries = 0
                    await self._consume()
            except websockets.WebSocketException as wse:
                retries += 1
                if retries > int(os.environ.get('APCA_RETRY_MAX', 3)):
                    await self.close()
                    self._running = False
                    raise ConnectionError("max retries exceeded")
                if retries > 1:
                    await asyncio.sleep(
                        int(os.environ.get('APCA_RETRY_WAIT', 3)))
                log.warn('websocket error, restarting connection: ' +
                         str(wse))
            finally:
                print(retries)
                if not self._running:
                    break
                await asyncio.sleep(0.01)`
Wouldn't this stop it from restarting on the next pass of the loop?

**Comment by nickvertucci** [comment]
@Chrisrdouglas Would this help solve 1006 websocket errors and reconnect/resubscribe?

**Comment by Chrisrdouglas** [comment]
@nickvertucci

hmm, I haven't looked at this code since march but I doubt it would solve the issue of getting disconnected because of that internal connection issue. Are a lot of people having that problem?

**Comment by noahkuhn** [comment]
I know this is a few months ago, but yes, I am having constant 1006 problems... During market hours I can't keep a stream running more than a few minutes. It throws a "Error in data transfer" error and then websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason

**Comment by srgustafson8** [comment]
I have been taking a look at this and while the cause of the 1006 errors remain a mystery to me, the reconnect should kick in after one and get you going again. 

I'm not a python expert, but @Chrisrdouglas assessment looks along the right lines. There was an update at some point since which brought `await self._consume()` out of the `if not self._running:` statement, but that doesn't seem to reconnect the websocket in the case of a 1006 error.  

I'll raise a pull request with a fix and see what the maintainers think.

**Comment by Chrisrdouglas** [comment]
@srgustafson8 Thank you for making a pull request with a fix. As for the 1006 error, I'm wondering if that has something to do with the way web sockets are implemented server side. If I hadn't already switched to Polygon for my needs then I'd use Wireshark to sniff the traffic on my machine and start trouble shooting the issue that way.

Alpaca really needs engineers who are testing and using the software they're providing :/

**Comment by srgustafson8** [comment]
I think I tracked down the cause of the 1006 error - at least when I was attempting to test my fix. I found a [stackoverflow thread talking about it](https://stackoverflow.com/a/58993145). As a caveat: i'm not a python dev or expert so could be wrong.

I could force a 1006 by pausing the thread and waiting for a bit. My understanding is that it's caused when a keepalive ping doesn't fire or respond, a combination of the ping settings on both client and server perhaps? 

I was likely getting it in my implementation due to my poor understanding of Python/asyncio, which meant that I was blocking the thread at some point in my code. so in addition to handling connection closed in this library i'll be trying to get my websocket consumer to run in a totally separate and isolated thread.
